packages/server-wallet/src/wallet/index.ts

```diff
@@ -0,0 +1,238 @@
+    throw 'Unimplemented';
+  }
+  async closeChannel(_channelId: Bytes32): Promise<ChannelResult> {
+    throw 'Unimplemented';
+  }
+  async getChannels(): Promise<ClientChannelResult[]> {
+    throw 'Unimplemented';
+  }
+
+  async pushMessage(message: AddressedMessage): Promise<{channelResults?: ChannelResult[]}> {
+    const channelIds: Bytes32[] = [];
+
+    try {
+      await Channel.transaction(async tx => {
+        for (const ss of message.signedStates || []) {
+          // We ignore unsigned states
+          if (!ss.signatures?.length) return;
+
+          const channelId = calculateChannelId(ss);
+          let channel = await Channel.query(tx)
+            .where('channelId', channelId)
+            .first();
+
+          if (!channel) {
+            const addresses = ss.participants.map(p => p.signingAddress);
+            const signingWallet = await SigningWallet.query(tx)
+              .whereIn('address', addresses)
+              .first();
+
+            if (!signingWallet) {
+              logger.error(
+                {
+                  knownWallets: await SigningWallet.query(tx).select(),
+                  addresses,
+                },
+                'Not in channel'
+              );
+              throw Error('Not in channel');
+            }
+
+            const {address: signingAddress} = signingWallet;
+            const cols: RequiredColumns = {
+              ...ss,
+              vars: [addHash(ss)],
+              signingAddress,
+            };
+
+            channel = Channel.fromJson(cols);
+            const {channelId} = await Channel.query(tx).insert(channel);
+            channelIds.push(channelId);
+          } else {
+            ss.signatures?.map(sig => channel.addState(ss, sig));
+            await Channel.query(tx).update(channel);
+            channelIds.push(channel.channelId);
+          }
+        }
+      });
+    } catch (err) {
+      logger.error({err}, 'Could not push message');
+      throw err;
+    }
+
+    const {outbox} = await protocolEngine(channelIds);
+
+    return {
+      channelResults: [
+        {
+          outbox,
+          appData: '',
+          appDefinition: '',
+          channelId: '',
+          challengeExpirationTime: 0,
+          status: 'funding',
+          participants: [],
+          allocations: [],
+          turnNum: 0,
+        },
+      ],
+    };
+  }
+  onNotification(_cb: (notice: Notification) => void): {unsubscribe: () => void} {
+    throw 'Unimplemented';
+  }
+}
+
+type ExecutionResult = {ids: Bytes32[]; outbox: Outgoing[]};
+const protocolEngine = async (
+  ids: Bytes32[],
+  outbox: Outgoing[] = []
+): Promise<ExecutionResult> => {
+  if (ids.length === 0) return {ids, outbox};
+
+  const [channelId, ...nextIds] = ids;
+
+  const channel = await Channel.forId(channelId, undefined);
+  const actions = await executionLoop(channel);
+  const todos = actions.filter(isInternal);
```

Comment:
I'm a bit unsure about this approach. It seems that it's pretty restrictive to have one piece of code output actions for another part to execute - e.g. it's hard to make later actions conditional on the outcome of previous actions. An alternative would be for `executionLoop` to return a list of `messages`, to add to the outbox, and `channelIds` to add to the list to process.

Reply:
Can a message be a transaction that needs to be sent to the chain?

Reply:
In what I wrote above, `message` means "protocol message" - the thing that the app receives via a `MessageQueued` and forwards on to the opponents.

I think the example of the transaction is a good example here. The code inside a workflow for submitting a transaction needs to be something like:
```ts
try {
  const transactionId = await chain.sendEthDeposit(channelId, amt, ...etc);
  await store.recordTransactionId(transactionId);
} catch e {
  await store.recordFailedTransaction(e);
}
```
This is because we now need to store things like `transactionId` in the store, in order for the protocol to be restartable.

The logic above would be difficult to write using actions, as it isn't a "fire-and-forget" scenario; the next step is dependent on the outcome of the previous step. I anticipate that a lot of the actions that we're taking will be like this.

Reply:
> I'm a bit unsure about this approach. It seems that it's pretty restrictive to have one piece of code output actions for another part to execute - e.g. it's hard to make later actions conditional on the outcome of previous actions. An alternative would be for executionLoop to return a list of messages to add to the outbox, and channelIds to add to the processing queue.

You could argue that you can make later actions conditional on earlier actions by calling `takeNextAction` again.

```
try {
  const transactionId = await chain.sendEthDeposit(channelId, amt, ...etc);
  await store.recordTransactionId(channelId, transactionId);
} catch e {
  await store.recordFailedTransaction(channelId, e); // or retry in some cases etc.
}
```

I would think of this as a single action: `SubmitTransaction`: every time we submit a transaction, we should record its hash.

It seems difficult to write a system where you can be sure it works the same way when you restart it, unless you design the system to have some properties. One property that would do this for us is:
- the `takeNextAction` function returns zero or one actions
- if it returns zero actions, we remove that channelID from the channels queue
- if it returns one actions, we do it, and call `takeNextAction` again

If we execute multiple actions within `takeNextAction`, how do we ensure that if the system crashes after executing the first, then the next time we run it, it would execute the second?

Reply:
I don't think grouping some sections of code into actions and running them in another process actually helps with that either though - it just makes the logic harder to follow. In the example above, what happens if the `SubmitTransaction` action fails in between `sendEthDeposit` and `recordTransactionId`?

In general it seems very hard to get the property that "wherever the system crashes, even mid-operation, it will be fine". I think the property we're going for is more that if you shut your wallet down during an idle period, and start it up later, and feed in any events you missed in the mean-time, then the protocols will continue to run.

I think a lot of the loop functions will be of the form "query some state, write some state, send some messages", but where the messages will need some information from back from the store after the write (e.g. waiting to see if the write was successful / including the id of a created guarantor channel etc.). It seems natural to write this in the loop in that form, rather than writing it as "query some state" -> "[action] write some state, send some messages", or as "query some state" -> "[action] write some state" -> "[next loop] read some state" -> "[action] send some messages".